### ISSUE TITLE
Add kubectl create secret --from-base64-literal and --random-value-for-key

### ISSUE THREAD

**Issue by ash2k** [kubernetes/kubectl#1654]
### What would you like to be added?

`kubectl create secret --from-literal=key=secret-value` is very handy. I have a use case where I'm creating a secret that needs to be binary (an encryption key), be of a certain size, and use all of those bits for entropy. I'm taking randomness from `/dev/urandom` but cannot use it as a literal value since it's binary.

I could put the secret material into a file and use `kubectl create secret --from-file=./file` but:

- Secret data hitting the disk is undesirable as it's potentially less secure.
- Secret is potentially exposed on disk, even if briefly.
- It's a hassle to clean up the file in the script, potentially leaving it behind, exposing it later.

Ideally I'd use something like `kubectl create secret --from-base64-literal="key=$(head -c 64 /dev/urandom | base64)"` to generate a value of 64 bytes, base64 encode it, pass to kubectl. It would decode it, persist into the cluster.

Yet another idea (can open a separate issue) is to have `--random-value-for-key=key=64`. It would generate a random value of length 64 and persist it into the `key` key. Even nicer and more secure (secret is not exposed on disk and via command line flags, where it can be seen by other processes) for bootstrapping a secret (from a script or console).

/sig cli

### Why is this needed?

Better scripting experience, more secure secret data handling.

**Comment by ak20102763** [comment]
@xuzhenglun I understand your concern, I am using `crypto/rand` package that generates high entropy cryptographic randomness, which can be suitable to generate keys, tokens. 

And here is more information about this flag:
--random-value-for-key=key=size
 Specify a key and size to generate random binary value from crypto/rand package (e.g., --random-value-for-key=key=64). The size should be a positive integer representing the number of bytes.

**Comment by Bhargav-manepalli** [comment]
@xuzhenglun , I've implemented the ` --from-base64-literal=<base64 encoded key>` flag in `kubectl`. With this flag, when creating a secret, the secret will be created with the encoded value. The logic is handled in kubectl, where the base64-encoded string from the CLI arguments is decoded and passed as the decoded key to the API server. The API server will then create an encoded secret for us. This ensures flexibility for users while maintaining proper handling of encoded secrets.

**Comment by xuzhenglun** [comment]
That’s really awesome, thanks for your contribution. Feel free to raise a PR and link with this issue when you are ready. @Bhargav-manepalli

/assign @Bhargav-manepalli
/unassign

**Comment by ak20102763** [comment]
Great work @Bhargav-manepalli, I have the same implementation for this other flag. Please raise a PR.

**Comment by Bhargav-manepalli** [comment]
Thanks @ak20102763 , @xuzhenglun 
I've submitted a pull request that implements the --from-base64-literal flag for kubectl create secret generic.
The implementation includes the new flag, associated logic, unit tests.
I'd appreciate your review and feedback.

**Comment by ardaguclu** [comment]
I think this should be triaged by sig-cli first to move forward. Because adding new flag requires strong justification.

**Comment by ardaguclu** [comment]
There is a workaround as stated in https://github.com/kubernetes/kubectl/issues/1654#issuecomment-2346265086

```
head -c 64 /dev/urandom | kubectl create secret genetic my-secret --from-file=/dev/stdin
```

and this workaround is also a reasonable and looks not hacky. 

We usually would consider adding more flags, if there is no other way around and in that case there is.

**Comment by Bhargav-manepalli** [comment]
Hi @xuzhenglun , I raised a PR. Any comments on this ?

**Comment by xuzhenglun** [comment]
> Added --from-base64-literal feature to secret kubernetes/kubernetes#127609

I'm afraid you have to give a good reason to convince sig-cli of the necessity of this parameter. AFAIK, for signle key with random data, reading from stdin is a fine solution. maybe mulitiple key with random value? But I wander is this a real world scenario? 🤔 💭

**Comment by ash2k** [comment]
It's a good workaround, but it only works if you need a secret with a single key/value.

**Comment by Bhargav-manepalli** [comment]
@ak20102763  It supports for multiple key:value 's as well. i mean multiple keys with encoded values. 
`kubectl create secret generic my-secrey --from-base64-literal=user=<encoded string> --from-base64-literal=pass=<encoded string>`

**Comment by ash2k** [comment]
I was talking about the workaround :)

**Comment by ak20102763** [comment]
@ash2k / @xuzhenglun , how about this workaround , this way we can have multiple key/value pairs in single secret

`kubectl create secret generic my-secret \
  --from-file=key1=<(head -c 64 /dev/urandom) \
  --from-file=key2=<(head -c 128 /dev/urandom) \
  --from-file=key3=<(head -c 64 /dev/urandom)
`

**Comment by ash2k** [comment]
@ak20102763 Sure. But why make the user do this vs have flags that do what's necessary? I think working with secrets is quite important and default kubectl experience should be fool-proof as much as possible (mostly to avoid leaking secret data). So in my mind it makes sense to provide a clean UX for this.

I think it's not about "possible to do" vs "impossible to do without the flag(s)". It's about helping the user to avoid leaking the secret data. This is way more important and should be prioritized. Most people will not know all these shell tricks (I didn't!), they probably don't work on Windows. Users will just put their secret data into a file(s), which is strictly less secure as it exposes it on disk.

Again, we have those specialized `kubectl create` commands for better UX, not because one cannot do the same thing without them. It's about UX, not possible/impossible.

Does this make sense? WDYT?

**Comment by ak20102763** [comment]
@ash2k I agree, we have `kubectl create` commands just to give better UX, and the reason I thought `—random-value-for-key` would be better option is its ease of use, user just have to give number of bytes he/she wants in the secret and it will be created. This way user doesn’t even need to know any specialised commands to create binary data, and it will be applicable across platforms. It’s is also more secure in the way it creates secret as it doesn’t need any data from the disk.

**Comment by Bhargav-manepalli** [comment]
@ash2k  , The test `pull-kubernetes-verify ` is failing with error `command "kubectl create secret generic": flag name "from-base64-literal" is invalid, long form of flag names can only contain lowercase characters or dash (must match ^[a-z]+[a-z\-]*$)` .  Because, currently it's only accepting lowercases and dashes. 
https://github.com/Bhargav-manepalli/kubernetes/blob/10d393a4f8f946753b7b2ed7e017e0a907e51e0f/staging/src/k8s.io/kubectl/pkg/cmd/util/sanity/cmd_sanity.go#L147

Either to change the flag-name or add numeric support to the flag ?

**Comment by ash2k** [comment]
@Bhargav-manepalli I'm not on sig-cli, cannot make any decisions. But in my personal opinion numbers should be allowed in flag names, I don't see why that would be bad in general.

Alternatively, what are the good names for the flag if not `--from-base64-literal`? 🤔

**Comment by Bhargav-manepalli** [comment]
> @Bhargav-manepalli I'm not on sig-cli, cannot make any decisions. But in my personal opinion numbers should be allowed in flag names, I don't see why that would be bad in general.
> 
> Alternatively, what are the good names for the flag if not `--from-base64-literal`? 🤔

Any suggestions on this ? @xuzhenglun

**Comment by xuzhenglun** [comment]
> Any suggestions on this ? @xuzhenglun

In my personal opinion, I think this new flag could bring user better experience indeed.  But I'm new and not on charge, cannot make decision too. 

Would you please help us to triage this?  @ardaguclu

**Comment by ardaguclu** [comment]
As I tried to state in here https://github.com/kubernetes/kubectl/issues/1654#issuecomment-2360657615, we are hesitant to add new flags and it is considered only for the cases where there is no workaround. Because usability of a flag may seem reasonable from one point of view but it may also lead to a bad user experience from another point of view (for example, numerous flags in one command, just like `run` has not good UX).

In that case there are workarounds for every possible case and I'd not prefer adding a new flag for an imperative command (i.e. `create secret`). But let's leave this issue as open for now.

**Comment by valid-var** [comment]
I vote for `--from-base64-literal` flag, as it can be real handy for passing any abnormal symbols, such as tabs, newlines, etc and basically any byte data

**Comment by solvingj** [comment]
@ardaguclu @mpuckett159 
I completely acknowledge and relate to the problems you mentioned about maintainability. I support declining the `--random` flag. 

With that said, to be clear, regarding `--from-base64-literal`, there is no actual equivalent workaround:
- The suggested workaround does not support defining multiple keys in a secret
- The /dev/stdin trick doesn't work on windows.  
  - I know most posix native people like to disregard windows use cases, but it's worth mentioning

Finally, regardless of secrurity, using base64 encoding is the only viable way of passing strings with arbitrary special characters through a shell to kubectl.  By declining this PR, it will remain impossible to use the `kubectl create secret` with secrets which contain special characters. 

I'm willing to believe I've got something wrong or I've missed some other nuanced workaround. If so, please feel free to explain. 

In any case, thanks for your contributions.
